chainerx_cc/chainerx/native/reduce.h

```diff
@@ -11,19 +11,84 @@ namespace chainerx {
+constexpr int64_t SerialLen = 8;
+
+template <typename In, typename ReductionImpl, int8_t InNdim, typename T, int64_t n>
+struct ExpandedPairwiseReduction {
+    static T run(const IndexableArray<const In, InNdim>& in, IndexIterator<InNdim>& it_in, ReductionImpl&& impl, int64_t& i_reduce) {
+        T accum = ExpandedPairwiseReduction<In, ReductionImpl, InNdim, T, n / 2>::run(in, it_in, impl, i_reduce);
+        impl.Reduce(ExpandedPairwiseReduction<In, ReductionImpl, InNdim, T, n / 2>::run(in, it_in, impl, i_reduce), accum);
+        return accum;
+    }
+};
+
+template <typename In, typename ReductionImpl, int8_t InNdim, typename T>
+struct ExpandedPairwiseReduction<In, ReductionImpl, InNdim, T, 1> {
+    static T run(const IndexableArray<const In, InNdim>& in, IndexIterator<InNdim>& it_in, ReductionImpl&& impl, int64_t& i_reduce) {
+        T accum = impl.MapIn(native_internal::StorageToDataType<const In>(in[it_in]), i_reduce);
+        ++it_in, ++i_reduce;
+        return accum;
+    }
+};
+
+template <typename In, typename ReductionImpl, int8_t InNdim, typename T>
+T PairwiseReduction(
+        const IndexableArray<const In, InNdim>& in, IndexIterator<InNdim>& it_in, ReductionImpl&& impl, int64_t reduce_len, T* tree_accum) {
+    int64_t i_reduce = 0;
+    T accum = impl.Identity();
+
+    while (i_reduce < reduce_len / ExpandLen * ExpandLen) {
+        // Invoke dynamic pairwise reduction if `i_reduce` is multiple of `SerialLen * ExpandLen`.
+        if (i_reduce != 0 && i_reduce % (SerialLen * ExpandLen) == 0) {
+            int i = 0;
+            for (int64_t k = i_reduce >> 1; k % (SerialLen * ExpandLen) == 0; k >>= 1, ++i) {
+                impl.Reduce(tree_accum[i], accum);
+            }
+            tree_accum[i] = accum;
+            accum = impl.Identity();
+        }
+        // This increments `i_reduce` by `ExpandLen`.
+        impl.Reduce(ExpandedPairwiseReduction<In, ReductionImpl, InNdim, T, ExpandLen>::run(in, it_in, impl, i_reduce), accum);
+    }
+
+    // Accumulate residuals.
+    while (i_reduce < reduce_len) {
+        impl.Reduce(impl.MapIn(native_internal::StorageToDataType<const In>(in[it_in]), i_reduce), accum);
+        ++it_in, ++i_reduce;
+    }
+
+    // Accumulate tree nodes.
+    int i = 0;
+    for (int64_t k = (reduce_len / ExpandLen - 1) / SerialLen; k > 0; k >>= 1, ++i) {
+        if (k & 1) {
+            impl.Reduce(tree_accum[i], accum);
+        }
+    }
+
+    return accum;
+}
+
+constexpr int log2(int64_t v) { return v == 1 ? 0 : log2(v >> 1) + 1; }
+
 template <typename In, typename Out, typename ReductionImpl, int8_t InNdim = kDynamicNdim, int8_t OutNdim = kDynamicNdim>
 void ReductionKernel(ReductionKernelArg<In, Out, InNdim, OutNdim> arg, ReductionImpl&& impl) {
     auto it_in = arg.in_indexer.It(0, arg.out_indexer.total_size());
+    int64_t reduce_len = arg.in_indexer.total_size() / arg.out_indexer.total_size();
+    decltype(impl.Identity()) tree_accum[63 - log2(ExpandLen * SerialLen)];
```

Comment:
Why not declaring in `PairwiseReduction`?

Reply:
Thanks, I moved the declaration.